lib/auth/accountrecovery.go

```diff
@@ -33,13 +37,113 @@ import (
 )
 
 // fakeRecoveryCodeHash is bcrypt hash for "fake-barbaz x 8".
 // This is a fake hash used to mitigate timing attacks against invalid usernames or if user does
 // exist but does not have recovery codes.
 var fakeRecoveryCodeHash = []byte(`$2a$10$c2.h4pF9AA25lbrWo6U0D.ZmnYpFDaNzN3weNNYNC3jAkYEX9kpzu`)
 
+// ErrMaxFailedRecoveryAttempts is a user friendly error message to notify user that recovery attempt
+// has been temporarily locked and an email has been sent.
+var ErrMaxFailedRecoveryAttempts = trace.AccessDenied("too many incorrect attempts, please check your email and try again later")
+
+// StartAccountRecovery implements AuthService.StartAccountRecovery.
+func (s *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccountRecoveryRequest) (types.UserToken, error) {
+	if err := s.isAccountRecoveryAllowed(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Only user's with email as their username can start recovery.
+	if _, err := mail.ParseAddress(req.GetUsername()); err != nil {
+		return nil, trace.BadParameter("only emails as usernames are allowed to recover their account")
+	}
+
+	if err := s.verifyCodeWithRecoveryLock(ctx, req.GetUsername(), req.GetRecoveryCode()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Remove any other existing tokens for this user before creating a token.
+	if err := s.deleteUserTokens(ctx, req.Username); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	token, err := s.createRecoveryToken(ctx, req.GetUsername(), UserTokenTypeRecoveryStart, req.GetRecoverType())
+	return token, trace.Wrap(err)
+}
+
+// verifyCodeWithRecoveryLock counts number of failed attempts at providing a valid recovery code.
+// After MaxAccountRecoveryAttempts, user is temporarily locked from further attempts at recovering and also
+// locked from logging in. Modeled after existing function WithUserLock.
+func (s *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
+	user, err := s.Identity.GetUser(username, false)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			// If user is not found, still authenticate. It should
+			// always return an error. This prevents username oracles and
+			// timing attacks.
+			return s.verifyRecoveryCode(ctx, username, recoveryCode)
+		}
+		return trace.Wrap(err)
+	}
+
+	status := user.GetStatus()
+	if status.IsLocked && status.RecoveryAttemptLockExpires.After(s.clock.Now().UTC()) {
+		log.Debugf("%v exceeds %v failed account recovery attempts, locked until %v",
+			user.GetName(), defaults.MaxAccountRecoveryAttempts, apiutils.HumanTimeFormat(status.RecoveryAttemptLockExpires))
+		return trace.AccessDenied("too many incorrect recovery attempts, please try again later")
+	}
+
+	fnErr := s.verifyRecoveryCode(ctx, username, recoveryCode)
+	if fnErr == nil {
+		return nil
+	}
+
+	// Do not lock user in case if DB is flaky or down.
+	if trace.IsConnectionProblem(fnErr) {
+		return trace.Wrap(fnErr)
+	}
```

Comment:
```suggestion
	switch {
	case fnErr == nil:
		return nil
	case trace.IsConnectionProblem(fnErr):
		// Do not lock user if DB is flaky or down.
		return trace.Wrap(fnErr)
	}
```

Reply:
leaving as is, next PR (step 2) i found myself extracting the `trace.IsConnection` and some later code for re-use in another function